wallet-new/src/Cardano/Wallet/API/V1/Types.hs

```diff
@@ -849,7 +852,50 @@ instance Arbitrary WalletAddress where
                               <*> arbitrary
                               <*> arbitrary
 
-type AccountIndex = Word32
+newtype AccountIndex = AccountIndex { getAccIndex :: Word32 }
+    deriving (Show, Eq, Ord, Generic)
+
+instance Bounded AccountIndex where
+    -- NOTE: minimum for hardened key. See https://iohk.myjetbrains.com/youtrack/issue/CO-309
+    minBound = AccountIndex 2147483648
+    maxBound = AccountIndex maxBound
+
+mkAccountIndex :: Word32 -> Either Text AccountIndex
+mkAccountIndex index | index >= getAccIndex minBound = Right $ AccountIndex index
+                     | otherwise = Left $ sformat
+                            ("mkAccountIndex: Account index should be in range ["%int%".."%int%"]")
+                            (getAccIndex minBound)
+                            (getAccIndex maxBound)
+
+instance ToJSON AccountIndex where
+    toJSON = toJSON . getAccIndex
+
+instance FromJSON AccountIndex where
+    parseJSON =
+        either (fail . toString) pure . mkAccountIndex <=< parseJSON
```

Comment:
Note that if we define an explicit `WalletError` for `mkAccountIndex`, it'd be better to use `sformat` instead of `toString` to leverage a `Buildable` instance and have a proper _diagnostic_!